### ISSUE TITLE
Theming

### ISSUE THREAD

**Issue by chrisvxd** [measuredco/puck#139]
Puck is being deployed in a wide variety of situations that require different themes. We should support theming to support this.

Some ideas:

* classNames for everything that can be customised
* style tags for everything that can be customised
* a discrete API exposing a limited subset of style overrides
* custom render functions to enable users to take ownership of specific parts of the application

Let's have a discussion and form this into a more fully fledged series of proposals.

Related #122

**Comment by Leweyse** [comment]
Hi @chrisvxd !

I definitely support using custom properties. A great example of a similar implementation is `shadcn` (reference: [CSS properties](https://ui.shadcn.com/docs/theming)).

I use Tailwind CSS and would love it if the theme customization could be easily extended to Custom Fields or Plugins in my projects.

**Comment by JakeSidSmith** [comment]
I think if you want to appease everyone you're probably going to need both inline `style` and `className` support.

And just to throw an extra complication into the works, there'll be people using [styled-components](https://styled-components.com/) that would like to be able to wrap your existing components instead of applying global styles. This might be able to be achieved in a similar way to how [react-markdown](https://www.npmjs.com/package/react-markdown#appendix-b-components) allows customizing elements....

If you had an interface that allowed overriding/extending the styles in one place (in addition to `style` and `className` on each component) it'd be incredibly powerful. E.g.

```tsx
noIdeaWhatToCallThis: {
  Button: (ExistingComponent) => ({className, props}) => (
    <ExistingComponent {...props} className={classNames(className, 'extra')} />
  ),
  // Styled component example
  Button: (ExistingComponent) => styled(ExistingComponent)`color: red;`
}
```

**Comment by chrisvxd** [comment]
@Leweyse also a big fan of custom properties, so there could be something here. This might be a good way of doing a relatively light and controlled custom theming layer.

@JakeSidSmith I think what you're suggesting is the "custom render functions" option combined with "classNames", which is full UI override.

This is definitely the most powerful option and would allow people to do whatever they want, but I have concerns around the trade-off in making the internals part of a public API. If you remove a component internally, you now have to issue a breaking change. I imagine the risk of this happening decreases as Puck becomes more stable.

**Comment by BleedingDev** [comment]
I agree with the simple solution with custom properties. They are well-known and supported and they can cover a lot of use-cases. Custom render functions would be for complete white-labelling, which is way to advanced (and heavily paid in solutions like Builder.io).

**Comment by chrisvxd** [comment]
Initial pass being worked on by @Leweyse (see #175)

**Comment by chrisvxd** [comment]
I think #224 may unlock this work, and should probably be a blocker here.

**Comment by BehnamKaraj** [comment]
Hi everybody,
until there is a theming option, here is the simplest workaround how to change the look and feel of your app:

1. Copy the index.css file from the **node_modules/@measured/puck/dist/** somewhere in your project:
e.g., **./assets/css/index.css**
2. Import it as usual in your .tsx/.jsx/.js file: **import "./assets/css/index.css";**
3. Adjust the styling to your needs

By the way, a big **thank you** to you for this library, it's easy to use and works perfectly. Great job
Cheerio!

**Comment by chrisvxd** [comment]
@DevOfManyThings we're more likely to continue with our own implementation. We have an approach, but just haven't managed to prioritise this.

**Comment by Zn4rK** [comment]
> We have an approach, but just haven't managed to prioritise this.

Out of curiosity, what's the approach? I have a solution in place in one of my projects where I have a babel transformer going over Puck's published source, and looking for css modules. When it finds one, it essentially modifies the object;

From something like:
```js
var styles_module_default18 = { "PuckPreview": "classNameForPuckPreview" };
```

To something like:
```js
import { getInjectedStyles } from './customStyles'; // contains tools for setting styles for a specific component.

var old_css_module_default18 = { "PuckPreview": "classNameForPuckPreview" };
var styles_module_default18 = getInjectedStyles('Puck/components/Preview', old_css_module_default18);
```

And then re-publishes Puck (privately). I can then consume it in my real project. It works well, and let's me customize the specific parts I wanted to customize. And I fully expect to throw away this stopgap, once something like this lands.

**Comment by codercatdev** [comment]
Upvoting this one 🥇

**Comment by austinm911** [comment]
As a tailwind v4 user, would love to learn more about how to integrate it!

**Comment by tonioloewald** [comment]
FWIW I think the answer is css-variables.

**Comment by greg-truewind** [comment]
do we need help on this one?

**Comment by chrisvxd** [comment]
Thanks for your patience all. We're not looking for support, and know how to do this, but we've held off implementing this until the Puck UI stabilises to avoid introducing a new vector for breaking API changes.

Things are improving on that front, but we're not quite there yet. Please continue to hold🙏

**Comment by Virrjet** [comment]
It would be nice to have the theme typed and passable as a prop to Puck. :)

**Comment by Albert0sans** [comment]
updates?

**Comment by oKelvinCosta** [comment]
## CSS variable naming: semantic/contextual tokens vs. color-based tokens

Hi! I´m loving using puck but the stylization is very basic/limited yet, doing a light/dark mode is really hard. For while I did the best what is possible. 

<img width="500"  alt="Image" src="https://github.com/user-attachments/assets/cb8f18ef-834f-4974-9a0b-3e71707456a4" />
<img width="500" alt="Image" src="https://github.com/user-attachments/assets/3a4ca3e6-4c7c-4b9d-aab1-b13412f1a556" />

Currently, Puck's UI uses color-based CSS variable names (e.g. `--red-500`, `--blue-300`) for styling various parts of the interface. This approach makes customization harder than it needs to be, because a single color token like `--red-500` ends up being used in multiple unrelated UI contexts — so changing it affects everything at once, with no way to target specific parts independently.

See here, the problem:
<img width="737" height="817" alt="Image" src="https://github.com/user-attachments/assets/738326c2-3bb0-4dce-8855-d6eb9135abd8" />

A more modular and maintainable approach would be to adopt **semantic/contextual variable names**, similar to what shadcn/ui does — where variables describe *where* or *what* they style, not just their color value. For example:

- `--puck-color-subtitle-editor` instead of `--red-500`
- `--puck-color-action-primary` instead of `--blue-500`
- `--puck-color-drag-handle` instead of a shared gray token

This way, consumers of the library can restyle individual parts of the UI without unintended side effects across the rest of the editor.

This pattern is well-established in design systems (shadcn/ui, Radix, MUI) and would significantly improve the developer experience for anyone trying to integrate Puck into a branded product.


One more addition, for the future. It's not possible to use tailwind on component blocks dynamically, due to the tailwind purge that removes it before the puck loads, or something similar. This could be considered in the future.
Styling with pure CSS is tedious and verbose.

I hope this example could help.

Thanks for the hard work.

**Comment by FedericoBonel** [comment]
Hey @oKelvinCosta!

Thanks for the feedback.

You are right that the current tokens and CSS properties we are defining are not intended for external use. We are actively working on this issue at the moment, and you can check out the progress on this branch: https://github.com/puckeditor/puck/tree/fede/puck-17-theming

You can also check out our acceptance criteria in this PR: #1678

What we are working on would allow you to change editor-wide (e.g. `--puck-color-surface`) and component-specific (e.g. `--puck-button-color-bg`) colors, spacing, border radiuses, and typography.

That would allow you to do something like what you are mentioning.

---

Regarding Tailwind, you can use Tailwind with Puck. We have a video about it [here](https://youtu.be/tDib6DCIcs8), an article [here](https://dev.to/puckeditor/how-to-build-a-react-page-builder-puck-and-tailwind-v40-5e3o), although it has been a while since this was published, so the setup might be different, and we know a lot of people who are using it.

To use it, you have to apply the same level of care you would on any other website. Basically, you [cannot use dynamic class names](https://tailwindcss.com/docs/detecting-classes-in-source-files#dynamic-class-names) like `p-{paddingProp}`. You need to add the full class name, such as `paddingProp === "sm" ? "p-6" : "p-[4px]"`, somewhere because [Tailwind adds a build step that scans your codebase, checks for any recognized class names in plain text, and includes them if they are present](https://tailwindcss.com/docs/detecting-classes-in-source-files).

This is not a limitation with Puck, but a Tailwind limitation, (or a feature, depending on how you see it).

The workarounds are adding [class name safelists with all the class names you need](https://tailwindcss.com/docs/detecting-classes-in-source-files#safelisting-specific-utilities), or using inline styles for fully dynamic styles like paddings, margins, etc.

I normally recommend these options in this order:

1. Inline styles for numeric values if you want to fully support any primitive value, such as paddings, margins, etc.
2. Accept things like t-shirt sizes or similar, such as `"m"`, `"xl"`, or `"xxl"`, and map them to full class names.
3. Use class name safelists.

**Comment by FedericoBonel** [comment]
Closed with #1705
